tests/0011-produce_batch.c

```diff
@@ -280,8 +285,250 @@ static void test_partitioner (void) {
+                 rd_kafka_name(rk));
+        topic_name = test_mk_topic_name("0011_per_message_flag", 1);
+        test_create_topic(topic_name, topic_num_partitions, 1);
+    
+        rkt = rd_kafka_topic_new(rk, topic_name,
+                                 topic_conf);
+        if (!rkt)
+                TEST_FAIL("Failed to create topic: %s\n",
+                          rd_strerror(errno));
+        
+        /* Create messages */
+        rkpartition_counts = calloc(sizeof(int), topic_num_partitions);
+        dr_partition_count = calloc(sizeof(int), topic_num_partitions);
+        rkmessages = calloc(sizeof(*rkmessages), msgcnt);
+        for (i = 0 ; i < msgcnt ; i++) {
+                int *msgidp = malloc(sizeof(*msgidp));
+                *msgidp = i;
+                rd_snprintf(msg, sizeof(msg), "%s:%s test message #%i",
+                            __FILE__, __FUNCTION__, i);
+                
+                rkmessages[i].payload = rd_strdup(msg);
+                rkmessages[i].len     = strlen(msg);
+                rkmessages[i]._private = msgidp;
+                rkmessages[i].partition = jitter(0, topic_num_partitions - 1);
+                rkpartition_counts[rkmessages[i].partition]++;
+        }
+        
+        r = rd_kafka_produce_batch(rkt, partition, RD_KAFKA_MSG_F_PARTITION | RD_KAFKA_MSG_F_FREE,
+                                   rkmessages, msgcnt);
+        
+        /* Scan through messages to check for errors. */
+        for (i = 0 ; i < msgcnt ; i++) {
+                if (rkmessages[i].err) {
+                        failcnt++;
+                        if (failcnt < 100)
+                                TEST_SAY("Message #%i failed: %s\n",
+                                         i,
+                                         rd_kafka_err2str(rkmessages[i].err));
+                }
+        }
+        
+        /* All messages should've been produced. */
+        if (r < msgcnt) {
+                TEST_SAY("Not all messages were accepted "
+                         "by produce_batch(): %i < %i\n", r, msgcnt);
+                if (msgcnt - r != failcnt)
+                        TEST_SAY("Discrepency between failed messages (%i) "
+                                 "and return value %i (%i - %i)\n",
+                                 failcnt, msgcnt - r, msgcnt, r);
+                TEST_FAIL("%i/%i messages failed\n", msgcnt - r, msgcnt);
+        }
+        
+        free(rkmessages);
+        TEST_SAY("Per-message partition: "
+                 "Produced %i messages, waiting for deliveries\n", r);
+        
+        msgcounter = msgcnt;
+        /* Wait for messages to be delivered */
+        test_wait_delivery(rk, &msgcounter);
+        
+        for (i = 0; i < topic_num_partitions; i++) {
+                if (dr_partition_count[i] != rkpartition_counts[i]) {
+                        TEST_FAIL("messages were not sent to designated partitions");
```

Comment:
print partition number, expected count and actual count, to help troubleshooting